### PR TITLE
Request based MVCC APIs

### PR DESCRIPTION
**Pull request by sticnarf**
The original txn requests are removed. Instead, I want to make all requests use the request based style.
There is little difference between raw api and mvcc apis (except that mvcc apis need to resolve lock), so I combine their request traits to `KvRequest`.

- [x] merge #97 first

### CODE REVIEW

File: src/kv_client/request.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+use std::sync::Arc;
+
+pub trait KvRequest: Sync + Send + 'static + Sized + Clone {
+    type Result;
+    type RpcRequest;
+    type RpcResponse: HasError + Clone + Send + 'static;
+    /// A single `KvRequest` can be divided into a number of RPC requests because the keys span
+    /// several regions or a single RPC request is too large. Most of the fields in these requests
+    /// share the same content while `Payload` is the part which differs among the requests.
+    type Payload;
```
Comment (nrc) [inline src/kv_client/request.rs]: I don't think `Payload` is a great name here - it is very generic and to me feels more like a value-like thing than a key-like thing. It is maybe better than `KeyType`, but I think we could still do better.

File: src/kv_client/request.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+};
+
+use futures::future::BoxFuture;
+use futures::prelude::*;
+use futures::stream::BoxStream;
+use grpcio::CallOption;
+use kvproto::kvrpcpb;
+use std::sync::Arc;
+
+pub trait KvRequest: Sync + Send + 'static + Sized + Clone {
```
Comment (nrc) [inline src/kv_client/request.rs]: Do we still need the `clone` bound?

File: src/kv_client/request.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                .boxed(),
+        )
+    }
+
+    fn store_stream<PdC: PdClient>(
+        &mut self,
+        pd_client: Arc<PdC>,
+    ) -> BoxStream<'static, Result<(Self::Payload, Store<PdC::KvClient>)>>;
+
+    fn into_request<KvC: KvClient>(
```
Comment (nrc) [inline src/kv_client/request.rs]: This should not be `into_` if it takes `self` by reference. Maybe `make_rpc_request`?

File: src/kv_client/request.rs (hunk 2 above)
Comment (sticnarf) [reply src/kv_client/request.rs]: Seems not. Let me remove it.

File: src/kv_client/request.rs (hunk 3 above)
Comment (sticnarf) [reply src/kv_client/request.rs]: Sounds good.

File: src/kv_client/request.rs (hunk 1 above)
Comment (sticnarf) [reply src/kv_client/request.rs]: I change it to `Entries`. Any better suggestions?

Comment (nrc) [reply src/kv_client/request.rs]: What are the possible instantiations? Key, Vec<Key>, BoundRange, and Vec<BoundRange>, are there any more?

Comment (sticnarf) [reply src/kv_client/request.rs]: `Vec<kvrpcpb::Mutations>`

Comment (nrc) [reply src/kv_client/request.rs]: Operand? KeyData? InputData? Data? (in decreasing order of preference)

Comment (sticnarf) [reply src/kv_client/request.rs]: It's changed to `KeyData`.

**Comment by mapleFU** [comment]
I have a naive question: maybe `MockDispatch` and `self.mock_dispatch` should be in `#[cfg(test)]`, or you should just use `Dispatch` trait.

**Comment by sticnarf** [comment]
> I have a naive question: maybe `MockDispatch` and `self.mock_dispatch` should be in `#[cfg(test)]`, or you should just use `Dispatch` trait.

Just because we are reusing `KvRequest::execute` function.

In fact, I don't think `MockDispatch` is very useful. To test the `execute` function, we can just create a mock request and use `MockKvClient`. Other trait functions in `KvRequest` should be tested separately. The whole process should be tested by an integration test. Then we don't need  `MockDispatch`. @nrc

**Comment by nrc** [comment]
MockDispatch is a trait which lets the mocks hook in a dummy dispatch before doing a proper dispatch. I've renamed it in https://github.com/tikv/client-rust/pull/101. The impls need to be cfg(test), but the trait itself shouldn't be.

I think it is useful, it lets us unit test the high-level parts of a request without the complications of rpc, network errors, retrying requests, etc.

**Pull request merged**